pkg/kv/kvserver/replica_range_lease.go

```diff
@@ -374,9 +372,12 @@ func (p *pendingLeaseRequest) requestLeaseAsync(
 			// return a NotLeaseHolderError rather than hanging, otherwise we could
 			// prevent the caller from nudging a different replica into acquiring the
 			// lease.
-			err := contextutil.RunWithTimeout(ctx, taskName, timeout, func(ctx context.Context) error {
-				return p.requestLease(ctx, nextLeaseHolder, reqLease, status, leaseReq)
-			})
+			//
+			// Does not use RunWithTimeout(), because we do not want to mask the
+			// NotLeaseHolderError on context cancellation.
+			requestLeaseCtx, requestLeaseCancel := context.WithTimeout(ctx, timeout) // nolint:context
+			defer requestLeaseCancel()
+			err := p.requestLease(requestLeaseCtx, nextLeaseHolder, reqLease, status, leaseReq)
```

Comment:
Did the outer code rely on this `err` being a `TimeoutError` in case of a timeout? Cause now it won't be anymore. Do you need to wrap it somewhere down this line, similarly to how `RunWithTimeout` [did](https://github.com/cockroachdb/cockroach/blob/9e44f568647bc06f2e0e33102e09df467f4b8f9a/pkg/util/contextutil/context.go#L93)?

Reply:
No, rather the opposite -- the upper-level code relied on it being a `NotLeaseHolderError`, which got masked by the `TimeoutError`.

Reply:
We want a lease acquisition timeout to result in a NLHE, so that the DistSender will go try some other replica that may be able to acquire the lease instead. We therefore explicitly construct a NLHE on context cancellation. In the case where the client context was cancelled (e.g. by a disconnect), the DistSender will bail out anyway, around here:

https://github.com/cockroachdb/cockroach/blob/f77e393230fbedb57e5caa0ebbf26356b3899c73/pkg/kv/kvclient/kvcoord/dist_sender.go#L2248-L2262

For reference, the DistSender is a fat KV RPC client that's responsible for e.g. routing KV requests to the correct ranges and leaseholders.